rclcpp/resource/logging.hpp.em

```diff
@@ -0,0 +1,101 @@
+// generated from rclcpp/resource/logging.hpp.em
+
+// Copyright 2017 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__LOGGING_HPP_
+#define RCLCPP__LOGGING_HPP_
+
+#include "rcutils/logging_macros.h"
+
+// These are used for compiling out logging macros lower than a minimum severity.
+#define RCLCPP_LOG_MIN_SEVERITY_DEBUG 0
+#define RCLCPP_LOG_MIN_SEVERITY_INFO 1
+#define RCLCPP_LOG_MIN_SEVERITY_WARN 2
+#define RCLCPP_LOG_MIN_SEVERITY_ERROR 3
+#define RCLCPP_LOG_MIN_SEVERITY_FATAL 4
+#define RCLCPP_LOG_MIN_SEVERITY_NONE 5
```

Comment:
Are the tests in this PR covering this ?

I'm surprised that these levels don't match the severity enums defined in rcutils, (https://github.com/ros2/rcutils/blob/a2f722530c5e39eecfeb62b8c40db99bf9ae2a5e/include/rcutils/logging.h#L123-L128) is it on purpose ?

Reply:
I think you've misinterpreted their purpose, they're for compiling out macros, not setting severity level of messages/loggers. My comment above is relevant here: "I have created RCLCPP-specific versions of e.g. `RCUTILS_LOG_MIN_SEVERITY_DEBUG` so that rclcpp macros can be compiled out independently to rcutils macros being compiled out (and can be completely compiled out once they do more work than just forwarding to rcutils)."

Because of the compile-time nature there are no tests for them, but I've tested it manually (that was how I discovered the need for https://github.com/ros2/rcutils/pull/60)

Reply:
I understood the purpose, the fact that we define different enums values for them still puzzles me though (and yeah I indeed missed  ros2/rcutils#60 which introduce the same "error" at the rcutils level IMO).
If a user want to add a custom severity, (let's say between debug and info for example), he can do it thanks to the values of rcutils:
`RCUTILS_LOG_SEVERITY_VERBOSE_MYSEVERITY = 15`.
 but cannot compile only those out because these enums don't leave any room between the levels. But now he cannot just compile out by using the same level he defined in rcutils, he has to find something between `RCLCPP_LOG_MIN_SEVERITY_DEBUG` and `RCLCPP_LOG_MIN_SEVERITY_INFO`.
How do we expect users to do this ? by defining:
```c
#define RCLCPP_LOG_MIN_SEVERITY_MYSEVERITY RCUTILS_LOG_SEVERITY_MYSEVERITY/10
```
or with hardcoded floating point values ?
```c
#define RCLCPP_LOG_MIN_SEVERITY_MYSEVERITY 1.5
```
It seems more user friendly to allow them to use the same values rather than a different one at each level, but maybe there is a technical constraint I missed here

Reply:
The reasoning was that, while a user can add their own severity level and call to the log _functions_ with their custom level, it's not straightforward for them to extend the logging _macros_ to have ones for their severity (even with the values matching the enum values). Since we need the min severity to be a preprocessor directive, we can't directly use the enum values (i.e. `#define RCLCPP_LOG_MIN_SEVERITY_MYSEVERITY RCUTILS_LOG_SEVERITY_MYSEVERITY` would not work for the purpose of compiling out the macros, which is what prompted https://github.com/ros2/rcutils/pull/60). So, simply to avoid the need for factoring out the _values_ used by the severity enum (10, 20, etc) into preprocessor directives, the values for the min severities are independent to the severity values themselves because there are other limiting factors on users creating custom severity macros anyway.

It could be done though, definitely.